src/psyclone/psyir/transformations/intrinsics/dotproduct2code_trans.py

```diff
@@ -0,0 +1,321 @@
+# -----------------------------------------------------------------------------
+# BSD 3-Clause License
+#
+# Copyright (c) 2022, Science and Technology Facilities Council
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# * Redistributions of source code must retain the above copyright notice, this
+#   list of conditions and the following disclaimer.
+#
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author: R. W. Ford, STFC Daresbury Lab
+
+'''Module providing a transformation from a PSyIR DOT_PRODUCT operator
+to PSyIR code. This could be useful if the DOT_PRODUCT operator is not
+supported by the back-end or if the performance in the inline code is
+better than the intrinsic.
+
+'''
+from __future__ import absolute_import
+from psyclone.psyir.nodes import BinaryOperation, Assignment, Reference, \
+    Loop, Literal, ArrayReference, Range, Routine
+from psyclone.psyir.symbols import DataSymbol, INTEGER_TYPE, REAL_TYPE, \
+    ArrayType, ScalarType
+from psyclone.psyir.transformations import TransformationError
+from psyclone.psyir.transformations.intrinsics.operator2code_trans import \
+    Operator2CodeTrans
+
+
+def _get_array_bound(vector1, vector2, writer):
```

Comment:
This is useful functionality that perhaps should be in `ArrayMixin` rather than here? Possibly it should be called something like `subsection_shape`? We have similar functionality when processing WHERE statements which all needs rationalising. (e.g. #717)

Reply:
In its current form it is specialised for this case (2 args and no validation). I've added this information to the docstring and referenced #717. That may or may not be enough. As for changing the name, at the moment it only works for full array dimension accesses (e.g. using (:)) so I don't think subsection shape is any better. I've left the name as is for the moment but please request it again if you disagree.